# Objetivos do Estudo

**Contexto: Qual é a situação que motivou essa análise?**

O coronavírus pegou o mundo todo de surpresa mudando a rotina das pessoas. Os moradores das cidades já não passavam mais seu tempo livre fora de casa, indo a cafés e shoppings; a maioria ficou em casa lendo livros. Isso chamou a atenção de startups que se apressaram para desenvolver novos aplicativos para os amantes de livros.

Recebemos um banco de dados de um dos serviços concorrentes nesse mercado. Ele contém dados sobre livros, editoras, autores, e classificação de clientes e avaliação de livros. Essa informação será usada para gerar uma proposição válida para o novo produto.  

**Proposições a senem analisadas:**
  
- Quantos livros foram lançados após 2000?

- Qual é o número de avaliações e a classificação média por livro?

- Qual editora lançou mais livros com mais de 50 páginas?

- Qual autor tem a maior média de classificação?

- Qual é a média de avaliações dos usuários mais ativos?

# Bibliotecas e funções usadas

## Bibliotecas:

In [1]:
import pandas as pd
from sqlalchemy import create_engine

## Funções:

In [2]:
def conexao(query: str):
    ''' 
    Essa função recebe uma query de uma consulta no banco SQL, e retorna o dataframe dessa consulta. 
    '''
    db_config = {
     'user': 'practicum_student', # username
     'pwd': 'QnmDH8Sc2TQLvy2G3Vvh7', # password
     'host': 'yp-trainers-practicum.cluster-czs0gxyx2d8w.us-east-1.rds.amazonaws.com',
     'port': 5432, # connection port
     'db': 'data-analyst-final-project-db' # the name of the database
     }
    
    connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(
        db_config['user'],
        db_config['pwd'],
        db_config['host'],
        db_config['port'],
        db_config['db'])
    
    engine = create_engine(connection_string, connect_args={'sslmode':'require'})

    return pd.io.sql.read_sql(query, con = engine)


# 1. Imprimindo as primeiras linhas das tabelas do banco de dados

In [3]:
# books:
query_books = '''
SELECT 
    *
FROM 
    books
LIMIT 
    5;
'''

books_5 = conexao(query_books)

books_5

,book_id,author_id,title,num_pages,publication_date,publisher_id
0,1,546,'Salem's Lot,594,2005-11-01,93
1,2,465,1 000 Places to See Before You Die,992,2003-05-22,336
2,3,407,13 Little Blue Envelopes (Little Blue Envelope...,322,2010-12-21,135
3,4,82,1491: New Revelations of the Americas Before C...,541,2006-10-10,309
4,5,125,1776,386,2006-07-04,268


In [4]:
# authors:
query_authors = '''
SELECT 
    *
FROM 
    authors
LIMIT 
    5;
'''

authors_5 = conexao(query_authors)

authors_5


,author_id,author
0,1,A.S. Byatt
1,2,Aesop/Laura Harris/Laura Gibbs
2,3,Agatha Christie
3,4,Alan Brennert
4,5,Alan Moore/David Lloyd


In [5]:
# publishers:
query_publishers = '''
SELECT 
    *
FROM 
    publishers
LIMIT 
    5;
'''

publishers_5 = conexao(query_publishers)

publishers_5


,publisher_id,publisher
0,1,Ace
1,2,Ace Book
2,3,Ace Books
3,4,Ace Hardcover
4,5,Addison Wesley Publishing Company


In [6]:
# ratings:
query_ratings = '''
SELECT 
    *
FROM 
    ratings
LIMIT 
    5;
'''

ratings_5 = conexao(query_ratings)

ratings_5


,rating_id,book_id,username,rating
0,1,1,ryanfranco,4
1,2,1,grantpatricia,2
2,3,1,brandtandrea,5
3,4,2,lorichen,3
4,5,2,mariokeller,2


In [7]:
# reviews:
query_reviews = '''
SELECT 
    *
FROM 
    reviews
LIMIT 
    5;
'''

reviews_5 = conexao(query_reviews)

reviews_5


,review_id,book_id,username,text
0,1,1,brandtandrea,Mention society tell send professor analysis. ...
1,2,1,ryanfranco,Foot glass pretty audience hit themselves. Amo...
2,3,2,lorichen,Listen treat keep worry. Miss husband tax but ...
3,4,3,johnsonamanda,Finally month interesting blue could nature cu...
4,5,3,scotttamara,Nation purpose heavy give wait song will. List...


# 2. Encontrando o número de livros lançados depois de 1 de janeiro de 2000.

In [8]:
# Precisamos escrever uma query em SQL para contar o número de livros lançados depois de '2000-01-01' e armazenar o resultado na coluna 'livros_apos_2000':
query_count = ''' 
SELECT 
    COUNT(book_id) AS livros_apos_2000
FROM 
    books
WHERE 
    CAST(publication_date AS date) > '2000-01-01';
'''

livros_apos_2000 = conexao(query_count)

livros_apos_2000

,livros_apos_2000
0,819


conclusão: 

Temos 819 livros lançados após 1 de janeiro de 2000

Recomendações:

O catálogo é majoritariamente moderno: Recomendo otimizar o estoque e o marketing para a Geração Z e Millennials, que dominam o consumo dessa era.

# 3. Encontrando o número de avaliações e a classificação média para cada livro.

In [9]:
query_rating_reviews = ''' 
SELECT 
    books.title AS livro,
    AVG(ratings.rating) AS avg_rating,
    COUNT(reviews.review_id) AS count_reviews
FROM 
    books
INNER JOIN
    ratings 
    ON ratings.book_id = books.book_id 
INNER JOIN
    reviews 
    ON reviews.book_id = ratings.book_id
GROUP BY
    books.title
ORDER BY
    avg_rating DESC,
    count_reviews DESC
'''

rating_and_reviews = conexao(query_rating_reviews)

rating_and_reviews.head()

,livro,avg_rating,count_reviews
0,A Dirty Job (Grim Reaper #1),5.0,16
1,School's Out—Forever (Maximum Ride #2),5.0,12
2,Moneyball: The Art of Winning an Unfair Game,5.0,9
3,In the Hand of the Goddess (Song of the Liones...,5.0,6
4,Neil Gaiman's Neverwhere,5.0,4


conclusão:

Os top 3 títulos com maior número de reviews e avaliação média são respectivamente:

- 1°: A Dirty Job (Grim Reaper #1)
- 2°: School's Out—Forever (Maximum Ride #2)
- 3°: Moneyball: The Art of Winning an Unfair Game

Recomendações:

Criar campanhas de venda casada (cross-selling) e destaque esses três títulos na home, pois eles são seus maiores ímãs de conversão e tráfego orgânico.

# 4. Identificando a editora que lançou o maior número de livros com mais de 50 páginas 

In [10]:
query_editoras = ''' 
SELECT 
    publishers.publisher as editora,
    COUNT(books.book_id) AS livros_com_mais_de_50_pg
FROM 
    books
INNER JOIN 
    publishers
    ON publishers.publisher_id = books.publisher_id
WHERE
    books.num_pages > 50
GROUP BY 
    publishers.publisher
ORDER BY
    livros_com_mais_de_50_pg DESC
'''

editoras_50 = conexao(query_editoras)
editoras_50.head()

,editora,livros_com_mais_de_50_pg
0,Penguin Books,42
1,Vintage,31
2,Grand Central Publishing,25
3,Penguin Classics,24
4,Ballantine Books,19


conclusão:

O top 3 editoras que lançaram o maior número de livros com mais de 50 páginas é:

- 1°: Penguin Books
- 2°: Vintage
- 3°: Grand Central Publishing

Recomendações: 

Estreitar o relacionamento comercial e negocie margens de lucro agressivas com Penguin, Vintage e Grand Central, já que elas garantem a volumetria do ecossistema.

# 5. Identificando o autor com a média mais alta de classificação de livros: Olharemos apenas para livros com pelo menos 50 classificações.

In [11]:
query_editoras = ''' 
SELECT 
    authors.author AS autor,
    AVG(books_50.avg_book_rating) AS avg_rating
FROM (-- Subquery selecionando apenas os livros com pelo menos 50 classificações
    SELECT 
        book_id,
        AVG(rating) AS avg_book_rating
    FROM ratings
    GROUP BY book_id
    HAVING COUNT(rating) >= 50
) AS books_50
INNER JOIN books 
    ON books.book_id = books_50.book_id
INNER JOIN authors 
    ON authors.author_id = books.author_id
GROUP BY 
    authors.author
ORDER BY 
    avg_rating DESC
'''

autores_50 = conexao(query_editoras)
autores_50.head()

,autor,avg_rating
0,J.K. Rowling/Mary GrandPré,4.283844
1,Markus Zusak/Cao Xuân Việt Khương,4.264151
2,J.R.R. Tolkien,4.258446
3,Louisa May Alcott,4.192308
4,Rick Riordan,4.080645


conclusão:

O pódio dos autores com a média mais alta de classificação de livros lançados, para livros com pelo menos 50 classificações é:

- 1°: J.K. Rowling/Mary GrandPré
- 2°: Markus Zusak/Cao Xuân Việt Khương
- 3°: J.R.R. Tolkien

Recomendações:

Garantir exclusividade ou campanhas de pré-venda para lançamentos desses autores (ou coautores), pois a base de fãs deles compra de olhos fechados.

# 6. Encontrando o número médio de avaliações entre usuários que classificaram mais do que 50 livros

In [12]:
query_usuarios = '''
SELECT 
    AVG(reviews_50.quantidade_de_avaliacoes) AS avg_reviews
FROM (-- Subquery para selecionar apenas os usuários que clasificaram mais de 50 livros
    SELECT
        username,
        COUNT(rating_id) As quantidade_de_classificacoes
    FROM 
        ratings
    GROUP BY 
        username
    HAVING 
        COUNT(rating_id) > 50
    ) AS ratings_50
INNER JOIN 
    (-- Subquery para contar a quantidade de avaliações por usuário
    SELECT
        username,
        COUNT(review_id) As quantidade_de_avaliacoes
    FROM 
        reviews
    GROUP BY 
        username
    ) AS reviews_50
    ON reviews_50.username = ratings_50.username
'''

# Executando a consulta
resultado_usuarios = conexao(query_usuarios)
resultado_usuarios

,avg_reviews
0,24.333333


conclusão: 

Os usuários mais ativos na plataforma (Aqueles que classificam mais de 50 livros) possuem um número médio de 24 avaliações.

Recomendações:

Criar um programa de fidelidade VIP ou gamificação que recompense o usuário ao atingir a marca de 24 avaliações, retendo os clientes que mais geram prova social gratuita para a plataforma.